src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlBuffer.cs

```diff
@@ -815,14 +815,17 @@ internal SqlGuid SqlGuid
                 }
                 else if (StorageType.SqlGuid == _type)
                 {
-                    return IsNull ? SqlGuid.Null : (SqlGuid)_object;
+                    return IsNull ? SqlGuid.Null : new SqlGuid(_value._guid);
                 }
                 return (SqlGuid)SqlValue; // anything else we haven't thought of goes through boxing.
             }
             set
             {
                 Debug.Assert(IsEmpty, "setting value a second time?");
-                _object = value;
+                if (!value.IsNull)
```

Comment:
If `_value` is already set and the new `value` is set to null, there might be an inconsistency where `_value` still points to a non-null value while `_isNull` indicates that it's null. This could lead to confusion and unexpected behavior in your code. It's crucial to ensure that `_value` and `_isNull` are consistent and accurately represent the state of the variable. IMO, If the intention is to set the variable to null, both `_value` and `_isNull` should be appropriately updated to reflect this.

Reply:
`_value` is a value type. it is the value it doesn't point to it. If the value is non-empty and then is updated to null it doesn't matter if there is a non-empty value in the variable because it won't be read through any of the code paths.

Reply:
The class definitely requires its own unit tests.
I have introduced the bug in this PR in [`SqlBuffer.get_SqlValue`](https://github.com/dotnet/SqlClient/pull/2306/files/54328bdbd429029273d0f48603a800f1c0317fd2#diff-48a7ee0a7ba0aea47e0131ff3033e85fe67de5120028484929bc0a8aad8f371cR988) 😥:
```csharp
case StorageType.SqlBinary:
case StorageType.SqlGuid:
    return _object;
```

Reply:
I don't understand what is the right place to put unit tests.
I thought it is 'src\Microsoft.Data.SqlClient\tests\FunctionalTests\Microsoft.Data.SqlClient.Tests.csproj' but this project does not target all targets of Microsoft.Data.SqlClient, particularly, it misses `net80`.
May you point me to the right direction? I want to fix the bug ASAP, @Wraith2 , @David-Engel , @DavoudEshtehari

Reply:
We don't have a net8 build currently. Write the test and use the correct attribute or preprocessor definition so that when the net8 build does exist (possibly soon-ish) it will be enabled and work.

Functional tests are run without a database connection. Anything that can be checked without running a query can go in there. Manual tests are run with a database connection, anything which can't be in functional tests belongs there.

Reply:
> We don't have a net8 build currently

Net8 and its related changes were merged last week by #2230

Reply:
I found and fixed a couple of bugs, but faced the issue with functional tests:
- I cannot find the right way to make internal classes of Microsoft.Data.SqlClient available for Microsoft.Data.SqlClient.Tests.
`[assembly: InternalsVisibleTo("FunctionalTests")]` does not work and, maybe, it is not the right way :\
Thought, @Wraith2 ?

PS: Also, it is very inconvenient that there is no way to run a particular unit test from IDE :(

Reply:
We should not use internals visible to. You'd need to use reflection if you want to test internals. Or, possibly include the file in the tests so it is local and then use it directly but this can be difficult because of dependencies in the code.

What is the root of the bugs and how did the get past previous review?

Reply:
Should the merge be reverted?

Reply:
@Wraith2 what approach is better (already used widely)?

@ErikEJ, I don't like revert commits and avoid them as many ch as possible. But it is up to you. FYI I'm preparing PR with the fixes and will submit it before Monday.

Reply:
It depends. I'd like more details on the bug and how it was found.
If the bug is a logical bug found by looking at the code but there is no way for users to reach that piece of code and that's why there are no covering tests then it's not a huge problem. In this case i'd say work forwards from here.
If the bug is a practical but found by hitting incorrect behaviour at runtime then we're clearly missing code coverage of a reachable path and it's a more important problem. In this case i'd say revert and retry.

Reply:
These are the bugs I'm talking about: https://github.com/dotnet/SqlClient/compare/main...Intrigma:SqlClient:sqlbuffer-bugs

Also, I made alternative performance/memory improvements, which works much better than #2306: https://github.com/dotnet/SqlClient/compare/main...Intrigma:SqlClient:sqlbuffer-improvements
I will share my benchmark project and its result later with PR. For now lets decide what we do with the bugs, @Wraith2

Reply:
I vote for a revert given that a release is planned for 24 Jan - just my 2 cents

Reply:
Here the PR to fix that: #2310

Reply:
@Wraith2 suggested to revert this PR too. See https://github.com/dotnet/SqlClient/pull/2310#discussion_r1461497741

Reply:
I just did a PR to revert.

Reply:
Now we just have to wait for the MS team to work their way through the email notifications, wonder what we've all been doing all weekend, and then approve it 😁